### PR TITLE
Add Support for VS2019 & Asynchronous Auto Loading

### PR DESCRIPTION
**Pull request by jbparker**
Hey @jgalentine007,

Adding support for VS2019 ended up being a little trickier since synchronous auto loading now causes quite a stink when loading extensions this way:

synchronous_autoload

I went down the rabbit hole or converting to AsyncPackage, and I think this landed in a good place that simultaneously does both the VS 2019 addition and loading asynchronously (the former was as easy as you indicated).

I've put commit comments throughout the PR to help with justification of any changes that aren't trivial in terms of intent.

Please let me know your thoughts - definitely okay to break this into two PRs to make this easier if that helps or anything else!

References: Issue #4

### CODE REVIEW

File: VSGitDiff/GitDiff.cs [hunk 1]
```diff
@@ -36,89 +41,51 @@ internal sealed class GitDiff
...
+        /// Adds our command handlers for menu (commands must exist in the command table file).
         /// </summary>
-        public static GitDiff Instance
+        /// <param name="package">Owner package, not null.</param>
+        public static async System.Threading.Tasks.Task InitializeAsync(AsyncPackage package)
         {
-            get;
-            private set;
-        }
+            await package.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (jbparker) [inline VSGitDiff/GitDiff.cs]: Choice to add the switch to the UI thread based on the following documents:

- [Migrate to AsyncPackage](https://github.com/Microsoft/VSSDK-Extensibility-Samples/tree/master/AsyncPackageMigration)
- [Asynchronous and multithreaded programming within VS using the JoinableTaskFactory](https://devblogs.microsoft.com/premier-developer/asynchronous-and-multithreaded-programming-within-vs-using-the-joinabletaskfactory/)
- [How to: Provide an asynchronous Visual Studio service](https://docs.microsoft.com/en-us/visualstudio/extensibility/how-to-provide-an-asynchronous-visual-studio-service?view=vs-2019)

In all the examples, the switch to the UI thread (calling `await package.JoinableTaskFactory.SwitchToMainThreadAsync()`) seems to be done before calls to `GetServiceAsync`.

For `GetGlobalService`, the second article seems to indicate that it definitely needs to be done here as well, so this one is included after the call as well:

> It is not always obvious whether some VS API you’re calling may require marshaling to the UI thread. Here is a list of types, members, or tips to help identify when you should explicitly switch to the UI thread yourself using SwitchToMainThreadAsync() before calling into VS:
> 
> 1. Package.GetGlobalService()
> 2. Casting an object to any other type (object to IVsHierarchy for example). If the object being cast is a native COM object, the cast may incur a call to IUnknown.QueryInterface, and will most likely require the UI thread.
> 3. Almost any call to an IVs* interface.

File: VSGitDiff/GitDiff.cs [hunk 2]
```diff
@@ -36,89 +41,51 @@ internal sealed class GitDiff
...
-        /// Adds our command handlers for menu (commands must exist in the command table file)
+        /// Adds a menu command for a given command ID, invocation handler, and status request handler.
         /// </summary>
-        /// <param name="package">Owner package, not null.</param>
-        private GitDiff(Package package)
+        /// <param name="commandService">The service used to add handlers for menu commands.</param>
+        /// <param name="commandID">The numeric command ID.</param>
+        /// <param name="invokeHandler">The event handler called to execute the command.</param>
+        /// <param name="beforeQueryStatusHandler">The event handler called when a client requests the command status.</param>
+        private static void AddMenuCommand(OleMenuCommandService commandService, int commandID, EventHandler invokeHandler, EventHandler beforeQueryStatusHandler)
```
Comment (jbparker) [inline VSGitDiff/GitDiff.cs]: Wraps up adding a menu command for the 5 places where we do this.

Note: passing `null` for any of the commands is supported by OleMenuCommand (source [here](https://searchcode.com/codesearch/view/28535169)).

File: VSGitDiff/GitDiffPackage.cs [hunk 3]
```diff
@@ -6,13 +6,13 @@
 
 namespace VSGitDiff
 {
-    [PackageRegistration(UseManagedResourcesOnly = true)]
+    [PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading = true)]
```
Comment (jbparker) [inline VSGitDiff/GitDiffPackage.cs]: Changes for this file to support asynchronous auto loading and use of JoinableTaskFactory.SwitchToMainThreadAsync in InitializeAsync is based on [this MS GitHub repo on migrating to AsyncPackage](https://github.com/Microsoft/VSSDK-Extensibility-Samples/tree/master/AsyncPackageMigration).

File: VSGitDiff/GitDiffPackage.cs [hunk 4]
```diff
@@ -37,14 +37,11 @@ public GitDiffPackage()
         /// Initialization of the package; this method is called right after the package is sited, so this is the place
         /// where you can put all the initialization code that rely on services provided by VisualStudio.
         /// </summary>
-        protected override void Initialize()
+        protected override async System.Threading.Tasks.Task InitializeAsync(System.Threading.CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
         {
-            // pass in SCC provider service
             try
             {
-                var scciProvider = GetService(typeof(IVsRegisterScciProvider)) as IVsGetScciProviderInterface;
```
Comment (jbparker) [inline VSGitDiff/GitDiffPackage.cs]: `IVsGetScciProviderInterface` service acquisition moved into `InitializeAsync`.

File: VSGitDiff/source.extension.vsixmanifest [hunk 5]
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <PackageManifest Version="2.0.0" xmlns="http://schemas.microsoft.com/developer/vsx-schema/2011" xmlns:d="http://schemas.microsoft.com/developer/vsx-schema-design/2011">
     <Metadata>
-        <Identity Id="VSGitDiff.Jonathan Galentine.d5d61ff8-f8ae-4a06-92be-b0bbedc36e2f" Version="1.2" Language="en-US" Publisher="Jonathan Galentine" />
+        <Identity Id="VSGitDiff.Jonathan Galentine.d5d61ff8-f8ae-4a06-92be-b0bbedc36e2f" Version="1.3" Language="en-US" Publisher="Jonathan Galentine" />
```
Comment (jbparker) [inline VSGitDiff/source.extension.vsixmanifest]: Changes here (including removal of now unnecessary MPF dependency) are [based off of recommendations in this MS dev blog article on upgrading extensions to VS 2019](https://devblogs.microsoft.com/visualstudio/how-to-upgrade-extensions-to-support-visual-studio-2019/).

Comment (jbparker) [inline VSGitDiff/source.extension.vsixmanifest:4]: Changes here (including removal of now unnecessary MPF dependency) are [based off of recommendations in this MS dev blog article on upgrading extensions to VS 2019](https://devblogs.microsoft.com/visualstudio/how-to-upgrade-extensions-to-support-visual-studio-2019/).

File: VSGitDiff/VSGitDiff.csproj [hunk 6]
```diff
@@ -126,10 +126,14 @@
     <Reference Include="Microsoft.Build.Framework" />
     <Reference Include="Microsoft.CSharp" />
     <Reference Include="Microsoft.TeamFoundation.Controls">
-      <HintPath>..\..\..\..\..\..\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer\Microsoft.TeamFoundation.Controls.dll</HintPath>
+      <HintPath Condition="Exists('..\..\..\..\..\..\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer')">..\..\..\..\..\..\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer\Microsoft.TeamFoundation.Controls.dll</HintPath>
```
Comment (jbparker) [inline VSGitDiff/VSGitDiff.csproj:129]: These conditional hint paths are to allow building the project in VS2017 or VS2019. 

Note: the last one on both references is my path for VS2019 - can definitely remove that one if you want.

File: VSGitDiff/VSGitDiff.csproj [hunk 7]
```diff
@@ -46,7 +46,7 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>VSGitDiff</RootNamespace>
     <AssemblyName>VSGitDiff</AssemblyName>
-    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
```
Comment (jbparker) [inline VSGitDiff/VSGitDiff.csproj:49]: This was necessary in order to resolve some build warnings.

File: VSGitDiff/GitDiffPackage.cs (hunk 4 above)
Comment (jbparker) [inline VSGitDiff/GitDiffPackage.cs:45]: This was moved into `GitDiff.InitializeAsync`.

File: VSGitDiff/GitDiff.cs [hunk 8]
```diff
@@ -422,10 +380,9 @@ private List<string> SelectedItemFilePaths(EnvDTE80.DTE2 dte)
         /// Returns the file paths of the selected files in the Team Explorer - Changes window.
         /// </summary>
         /// <returns></returns>
-        private List<string> SelectedSCCFilePaths()
+        private static List<string> SelectedSCCFilePaths()
         {
             List<string> paths = new List<string>();                                
-            var teamExplorer = this.ServiceProvider.GetService(typeof(ITeamExplorer)) as ITeamExplorer;
```
Comment (jbparker) [inline VSGitDiff/GitDiff.cs:428]: This was moved into `InitializeAsync`.

File: VSGitDiff/GitDiff.cs [hunk 9]
```diff
@@ -383,20 +350,11 @@ private bool GitIsSCC()
                 return false;
         }
 
-        /// <summary>
-        /// Returns a static DTE2 debugging object.
-        /// </summary>
-        /// <returns>DTE2 debugging object.</returns>
-        private static EnvDTE80.DTE2 GetDTE2()
-        {
-            return Package.GetGlobalService(typeof(DTE)) as EnvDTE80.DTE2;
```
Comment (jbparker) [inline VSGitDiff/GitDiff.cs:392]: Also moved to `InitializeAsync`.

File: VSGitDiff/GitDiff.cs [hunk 10]
```diff
@@ -36,89 +41,51 @@ internal sealed class GitDiff
...
-        /// Adds our command handlers for menu (commands must exist in the command table file)
+        /// Adds a menu command for a given command ID, invocation handler, and status request handler.
         /// </summary>
-        /// <param name="package">Owner package, not null.</param>
-        private GitDiff(Package package)
+        /// <param name="commandService">The service used to add handlers for menu commands.</param>
+        /// <param name="commandID">The numeric command ID.</param>
+        /// <param name="invokeHandler">The event handler called to execute the command.</param>
+        /// <param name="beforeQueryStatusHandler">The event handler called when a client requests the command status.</param>
+        private static void AddMenuCommand(OleMenuCommandService commandService, int commandID, EventHandler invokeHandler, EventHandler beforeQueryStatusHandler)
```
Comment (jbparker) [inline VSGitDiff/GitDiff.cs:76]: Bundle up these calls into one function for clarity.

Note: passing null for the `changeHandler` (and sometimes, `beforeQueryStatusHandler`) is allowed in `OleMenuCommand` and how the different constructors are doing this internal to the class.

File: VSGitDiff/GitDiff.cs [hunk 11]
```diff
@@ -36,89 +41,51 @@ internal sealed class GitDiff
...
-        private readonly Package package;
-
-        /// <summary>
-        /// Gets the instance of the command.
+        /// Initializes static services of the command.
+        /// Adds our command handlers for menu (commands must exist in the command table file).
         /// </summary>
-        public static GitDiff Instance
+        /// <param name="package">Owner package, not null.</param>
+        public static async System.Threading.Tasks.Task InitializeAsync(AsyncPackage package, System.Threading.CancellationToken cancellationToken)
```
Comment (jbparker) [inline VSGitDiff/GitDiff.cs:48]: Overall placement of `JoinableTaskFactory.SwitchToMainThreadAsync` (which switches to UI thread) based on recommendations from the following MS articles:

- [Migrate to AsyncPackage](https://github.com/Microsoft/VSSDK-Extensibility-Samples/tree/master/AsyncPackageMigration)
- [Asynchronous and multithreaded programming within VS using the JoinableTaskFactory](https://devblogs.microsoft.com/premier-developer/asynchronous-and-multithreaded-programming-within-vs-using-the-joinabletaskfactory/)

All of the examples they pose seem to put that call before any calls to `GetServiceAsync`.

Additionally, the second article indicates that `GetGlobalService` should be called within this context as well:

> It is not always obvious whether some VS API you’re calling may require marshaling to the UI thread. Here is a list of types, members, or tips to help identify when you should explicitly switch to the UI thread yourself using SwitchToMainThreadAsync() before calling into VS:
> 
> 1. **Package.GetGlobalService()**
> 2. Casting an object to any other type (object to IVsHierarchy for example). If the object being cast is a native COM object, the cast may incur a call to IUnknown.QueryInterface, and will most likely require the UI thread.
> 3. Almost any call to an IVs* interface.

**Comment by jgalentine007** [comment]
Great work, thanks for the contribution! Will update on NuGet / Marketplace.

**Comment by jbparker** [comment]
My pleasure! Thanks for the quick review!

**Pull request merged**